### ISSUE TITLE
Artifact storage backends

### ISSUE THREAD

**Issue by Crevil** [lunarway/release-manager#179]
We are currently looking into extracting the artifact storage mechanism from the cluster configuration repository.

Currently all artifacts in CI are pushed directly to the cluster configuration repository with the `artifact` CLI. During busy hours we will have many builds trying to push its artifacts into the repository leading to a lot of waiting during congestion of pushes in the Git tree.

Hopefully we will be able to increase performance of the cluster configuration repository by reducing the number of commits to it. At Lunar we currently have around 44k commits in the repo where roughly 50% are artifact pushes.

Our current goal is to extract the storage into a blob storage like AWS S3 so pushes from CI can be done in parallel.

Artifacts will still be stored in a similar to what we do now in Git.

```
- <svc>
  - <branch>
    - <artifact-id>
      - <env>
        - deployment.yaml
      - artifact.json
- ...
- product
  - master
    - master-123456-789
      - prod
        - deployment.yaml
      - artifact.json
    - master-654321-987
      - ...
      - artifact.json
```

The interaction needed from the currently implemented flows are something like below.
This is ongoing and taken from branch `refactor/storage` work on in PR https://github.com/lunarway/release-manager/pull/178.

```go
type Storage interface {
	ArtifactExists(ctx context.Context, artifactID string) (bool, error)
	GetBranch(ctx context.Context, service, artifactID string) (string, error)

	LatestArtifactSpecification(context.Context, ArtifactLocation) (artifact.Spec, error)
	LatestArtifactPaths(ctx context.Context, service, environment, branch string) (specPath, resourcesPath string, close CloseFunc, err error)
	ArtifactSpecifications(ctx context.Context, service string, count int) ([]artifact.Spec, error)
	ArtifactPaths(ctx context.Context, service, environment, branch, artifactID string) (specPath, resourcesPath string, close CloseFunc, err error)
}
```

If anyone has input on the current artifact implementation that could make it easier to interact with please comment on this issue.

`ArtifactPaths` is responsible for fetching all resources related to an artifact and store it on the filesystem where it can be used.
I think we should make use of a better interface that the file system at some point, but for now this is a detail I'm not going to concert my self with.

**Comment by Crevil** [comment]
I've managed to boil the interface further down. It is beginning to look promising I think.

```go
type Storage interface {
	ArtifactExists(ctx context.Context, artifactID string) (bool, error)

	ArtifactSpecification(ctx context.Context, service, artifactID string) (artifact.Spec, error)
	ArtifactPaths(ctx context.Context, service, environment, branch, artifactID string) (specPath, resourcesPath string, close CloseFunc, err error)

	LatestArtifactSpecification(context.Context, ArtifactLocation) (artifact.Spec, error)
	LatestArtifactPaths(ctx context.Context, service, environment, branch string) (specPath, resourcesPath string, close CloseFunc, err error)

	ArtifactSpecifications(ctx context.Context, service string, count int) ([]artifact.Spec, error)
}
```

**Comment by Crevil** [comment]
Implemented in #183
